### DISCUSSION TITLE
About: semantically corresponding text

### DISCUSSION THREAD

**Discussion by kchanqvq** [dalanicolai/sketch-mode#3] [Ideas]
was: https://lists.gnu.org/archive/html/emacs-devel/2021-08/msg01554.html

Some possible advantage:
- better integration with other part of emacs
   + cursor movement commands
   + kill/yank
   + `isearch`
   + toggle between text representation/graphic representation (if the text representation is "isomorphic")
- integration with my own package crdt.el, to support collaborative sketching

In the spirit of having an "isomorphic" text representation,
I suggest directly using the svg s-exp as underlying text, but omitted the parts that isn't "really" part of the sketch "document" but merely graphical guides:
- grids
- cursor/selection indicator, for use with Emacs native movement/selection commands
- maybe in the future other temporary messages/annotations...

When using svg s-exp directly, `*ward-char` `*ward-word` doesn't seem to make sense, but we can easily
add advice to them and let them move in the unit of s-exp when inside a sketch overlay. We can let them move between elements and groups etc.

The isomorphic representation can be naturally implemented by an architecture where edit and render is separated (edit happens on underlying text, render merely create an svg from the text). IDK how much change it require for current implementation, but it sounds possibly cleaner in the long term.

**Comment by dalanicolai** [comment]
Haha, that's fast. Okay, so I still do not see the big advantage, although the collaborative sketching sounds fun. And there might be useful advantages that I just do not see yet.   Anyway, implementing it as you suggest should not do any harm, and I guess that implementing the change requires only a minor modification (i.e. would be very quick and easy to do).

You could simply fork it and create a branch for it. As I mentioned already, the code is very accessible (especially if you use edebug) and you could probably implement it in no time...

**Comment by dalanicolai** [comment]
Oops, I actually forgot that I did not create the code to create the overlay (I was confused by some pdf-reader experiments that I did, where I did create the code for creating the overlay myself). So in this package I simply use the `insert-image` function.
But then still I guess modifying it to make it behave as you suggest should not be too difficult (but maybe a little less easy than I assumed in my previous comment)

**Comment by dalanicolai** [comment]
Ah... okay... so it seems extremely easy because `insert-image` simply allows for passing a string as argument for the underlying text. And the image definition at that point is of course already available...
